modules/core/src/main/scala/gem/Location.scala

```diff
@@ -10,46 +11,167 @@ import scalaz.Ordering.{EQ, GT, LT}
+      0
+
+    override def toString: String =
+      "{α}"
+  }
+
+  /** A Location that falls somewhere between the Beginning and End.
+    */
+  sealed abstract case class Middle(posList: OneAnd[IList, Int]) extends Location {
+    def positions: Stream[Int] =
+      posList.toStream #::: Stream.continually(Int.MinValue)
+
+    def toIList: IList[Int] =
+      posList.head +: posList.tail
+
+    def toList: List[Int] =
+      posList.head :: posList.tail.toList
+
+    protected def minPrefixLength: Int =
+      1 + posList.tail.length
+
+    override def toString =
+      toList.mkString("{", ",", "}")
+  }
+
+  /** A marker for the last Location.  No other Location comes after the End.
+    * End should not be used to order elements unless you are sure no elements
+    * will ever need to be inserted after End.  Use End to calculate new
+    * Locations that fall after the last existing Location (if any).
+    */
+  case object End extends Location {
+    def positions: Stream[Int] =
+      Stream.continually(Int.MaxValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{ω}"
+  }
+
+  // Constructors
 
   def apply(is: Int*): Location =
-    new Location(trim(is)) {}
+    fromTrimmedIList(IList(is: _*))
+
+  def fromFoldable[F[_]: Foldable](fi: F[Int]): Location =
+    fromTrimmedIList(IList.fromFoldable(fi))
+
+  private def fromTrimmedIList(is: IList[Int]): Location =
+    is.dropRightWhile(_ === Int.MinValue) match {
+      case INil()      => Beginning
+      case ICons(h, t) => new Location.Middle(OneAnd(h, t)) {}
+    }
+
+  /** Assuming not all provided Ints are `Int.MinValue`, produces a `Middle`
+    * `Location`.
+    */
+  def unsafeMiddle(i: Int, is: Int*): Location.Middle =
+    unsafeMiddle(i :: is.toList)
+
+  /** Assuming not all provided Ints are `Int.MinValue`, produces a `Middle`
+    * `Location`.
+    */
+  def unsafeMiddle[F[_]: Foldable](fi: F[Int]): Location.Middle =
+    fromFoldable(fi) match {
+      case m: Location.Middle => m
+      case _                  => sys.error("Location arguments do not form a Middle position: " + fi.toList.mkString(", "))
+    }
 
-  def fromList(is: List[Int]): Location =
-    new Location(trim(is)) {}
 
-  def fromIList(is: IList[Int]): Location =
-    new Location(trim(is)) {}
+  // Utility
 
-  def fromNel(n: NonEmptyList[Int]): Location =
-    new Location(trim(n.list)) {}
+  /** Finds `count` `Location`s that fall evenly distributed between `l0` and
+    * `l1`, assuming these locations are not the same.
+    *
+    * @param count how many locations to find
+    * @param start starting location (exclusive)
+    * @param end ending Location (exclusive)
+    *
+    * @return sorted list of `Location` where every element is GT l0 and LT l1
+    *         (or vice versa if l0 is GT l1)
+    */
+  def find(count: Int, start: Location, end: Location): IList[Location.Middle] = {
+    val Zero  = BigInt(0)
```

Comment:
Right, it allows integer math without worrying about overflow.  The series of position integers can be thought of as digits in a number system with radix the entire range of `Int` values.